lib/client/rpc/market-clients/ammMarkets.ts

```diff
@@ -78,85 +75,146 @@ export class FutarchyAmmMarketsRPCClient implements FutarchyAmmMarketsClient {
     }
   }
 
+  validateAddLiquidity(
+    ammMarket: AmmMarket,
+    quoteAmount: number,
+    maxBaseAmount: number,
+    slippage: number
+  ): LiquidityAddError | null {
+    const quoteAmountArg = new BN(
+      quoteAmount *
+        new BN(10).pow(new BN(ammMarket.quoteToken.decimals)).toNumber()
+    );
+    const baseAmountArg = new BN(
+      maxBaseAmount *
+        new BN(10).pow(new BN(ammMarket.baseToken.decimals)).toNumber()
+    );
+
+    const baseAmountWithSlippage = calculateMaxWithSlippage(
+      baseAmountArg,
+      slippage
+    );
+    // base passed in should be ammBaseAmount honestly...
+    const baseReserve = ammMarket.baseAmount;
+    const quoteReserve = ammMarket.quoteAmount;
+
+    const ammBaseAmount = quoteAmountArg
+      .mul(baseReserve)
+      .div(quoteReserve)
+      .add(new BN(1));
+    if (baseAmountWithSlippage < ammBaseAmount.toNumber()) {
+      console.warn(
+        `liquidity max base exceeded. baseAmountArg: ${baseAmountWithSlippage}. quoteAmountArg: ${quoteAmountArg.toNumber()}, ammBaseAmount: ${ammBaseAmount.toNumber()}`
+      );
+      return "AddLiquidityMaxBaseExceeded";
+    }
+    return null;
+  }
+
   async addLiquidity(
-    ammAddr: PublicKey,
-    quoteAmount?: number,
-    baseAmount?: number
-  ) {
+    ammMarket: AmmMarket,
+    quoteAmount: number,
+    maxBaseAmount: number,
+    slippage: number
+  ): Promise<string[] | LiquidityAddError> {
     if (!this.transactionSender) return [];
 
-    const ammAcount = await this.ammClient.getAmm(ammAddr);
-    const minLpTokens = new BN(0);
+    const validationError = this.validateAddLiquidity(
+      ammMarket,
+      quoteAmount,
+      maxBaseAmount,
+      slippage
+    );
+    if (validationError) {
+      return validationError;
+    }
 
-    const quoteInfo = await this.rpcProvider.connection.getAccountInfo(
-      ammAcount.quoteMint
+    const quoteAmountArg = new BN(
+      quoteAmount *
+        new BN(10).pow(new BN(ammMarket.quoteToken.decimals)).toNumber()
     );
-    const quoteDecimals = unpackMint(ammAcount.quoteMint, quoteInfo).decimals;
-    const quoteAmountCasted = new BN(quoteAmount).mul(
-      new BN(10).pow(new BN(quoteDecimals))
+    const maxBaseAmountArg = new BN(
+      maxBaseAmount *
+        new BN(10).pow(new BN(ammMarket.baseToken.decimals)).toNumber()
     );
 
-    const baseInfo = await this.rpcProvider.connection.getAccountInfo(
-      ammAcount.baseMint
+    const quoteAmountWithSlippage = calculateMaxWithSlippage(
```

Comment:
So this isn't correct. Quote amount doesn't have slippage, just a future todo to fix.

Reply:
maxBaseAmount: BN,
minLpTokens: BN,

Are the two that require slippage calcualtion.

Reply:
I believe this is the issue https://github.com/metaDAOproject/metadao-frontend/issues/60